cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java

```diff
@@ -196,4 +230,80 @@ private String getProgramName(ProgramId programId) {
     return String.join(",", programId.getApplication(), programId.getType().getCategoryName(),
                        programId.getProgram());
   }
+
+  private long upsert(String namespace, String type, Config config) throws IOException {
+    List<Field<?>> primaryKey = getPrimaryKey(namespace, type, config.getName());
+    Optional<StructuredRow> row = table.read(primaryKey);
+    long seq = 1;
+    if (row.isPresent()) {
+      Long currentSeq = row.get().getLong(StoreDefinition.PreferencesStore.SEQUENCE_ID_FIELD);
+      if (currentSeq != null) {
+        seq = currentSeq + 1;
+      }
+    }
+    table.upsert(toFields(namespace, type, config, seq));
+    return seq;
+  }
+
+  private long delete(String namespace, String type, String name) throws IOException {
+    List<Field<?>> primaryKey = getPrimaryKey(namespace, type, name);
+    Optional<StructuredRow> row = table.read(primaryKey);
+    if (!row.isPresent() || row.get().getString(StoreDefinition.PreferencesStore.PROPERTIES_FIELD) == null) {
+      return -1;
+    }
+    Long currentSeq = row.get().getLong(StoreDefinition.PreferencesStore.SEQUENCE_ID_FIELD);
+    long seq = currentSeq == null ? 1L : currentSeq + 1;
+    table.upsert(toFields(namespace, type, name, seq));
+    return seq;
+  }
+
+  private void checkSeqId(String namespace, String type, String name, long seq)
+    throws IOException, ConflictException {
+    List<Field<?>> primaryKey = getPrimaryKey(namespace, type, name);
+    Optional<StructuredRow> row = table.read(primaryKey);
```

Comment:
You can also do it in functional way if preferred.

```java
    table.read(primaryKey)
      .map(r -> r.getLong(StoreDefinition.PreferencesStore.SEQUENCE_ID_FIELD))
      .filter(Objects::nonNull)
      .filter(currentSeq -> currentSeq >= seq)
      .orElseThrow(() -> new ConflictException(
        String.format("Expected sequence id >= %d for %s %s in namespace %s, but found %s",
                      seq, type, name, namespace, String.valueOf(currentSeq))))
```

Reply:
I know... I had it like that at first, I just think in this case, the non-functional style is easier to read, so I changed it.